src/Core/Animation/KeyFramedValueInterpolators.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef RADIUMENGINE_KEYFRAMEDVALUE_INTERPOLATOR_HPP
+#define RADIUMENGINE_KEYFRAMEDVALUE_INTERPOLATOR_HPP
+
+#include <Core/Animation/KeyFramedValue.hpp>
+#include <Core/Animation/Pose.hpp>
+#include <Core/Animation/PoseOperation.hpp>
+#include <Core/Math/Interpolation.hpp>
+
+namespace Ra {
+namespace Core {
+namespace Animation {
+
+/**
+ * Look for the keyframes around time \p t.
+ * \param[in] keyframes The list of keyframes.
+ * \param[in] t The time to search for.
+ * \param[out] i0 The index for the keyframe preceding t.
+ * \param[out] i1 The index for the keyframe following t.
+ * \param[out] dt Linear parameter, \f$ \in [0;1] \f$, of \p t between times \p t0
+ *             and \p t1 at keyframes i0 and i1.
+ * \note If \p t is lower than the first keyframe's time tf,
+ *       then \p i0 = \p i1 = 0 and \p dt = 0.
+ *       If \p t is higher than the last keyframe's time tl,
+ *       then \p i0 = \p i1 = the index of the last keyframe and \p dt = 0.
+ *       If \p t is an exact match on a keyframe's time tt,
+ *       then \p i0 = \p i1 = the index of the keyframe and \p dt = 0.
+ */
+template <typename T>
+void findRange( const typename KeyFramedValue<T>::KeyFrames& keyframes,
+                const Scalar& t,
+                size_t& i0,
+                size_t& i1,
+                Scalar& dt ) {
+    // before first
+    if ( t < keyframes.begin()->first )
+    {
+        i0 = 0;
+        i1 = i0;
+        dt = 0_ra;
+        return;
+    }
+    // after last
+    if ( t > keyframes.rbegin()->first )
+    {
+        i0 = keyframes.size() - 1;
+        i1 = i0;
+        dt = 0_ra;
+        return;
+    }
+    auto it = std::find_if( keyframes.begin(), keyframes.end(), [t]( const auto& kf ) {
+        return Math::areApproxEqual( kf.first, t );
+    } );
+    // exact match
+    if ( it != keyframes.end() )
+    {
+        i0 = std::distance( keyframes.begin(), it );
+        i1 = i0;
+        dt = 0_ra;
+        return;
+    }
+    // in-between
+    auto kf0   = keyframes[0];
+    kf0.first  = t;
+    auto upper = std::upper_bound(
+        keyframes.begin(), keyframes.end(), kf0, []( const auto& a, const auto& b ) {
+            return a.first < b.first;
+        } );
+    auto lower = upper;
+    --lower;
+    i0        = std::distance( keyframes.begin(), lower );
+    i1        = std::distance( keyframes.begin(), upper );
+    Scalar t0 = lower->first;
+    Scalar t1 = upper->first;
+    dt        = ( t - t0 ) / ( t1 - t0 );
+}
+
+/** \name Predefined KeyFramedValue interpolators.
+ * Linearly interpolates the keyframes right before and right after time \p t.
+ * \note If there is no keyframe before \p t, then returns the first keyframe.
+ * \note If there is no keyframe after \p t, then returns the last keyframe.
+ */
+/// \{
+template <typename T>
+inline T linearInterpolate( const typename KeyFramedValue<T>::KeyFrames& keyframes, Scalar t ) {
```

Comment:
This seems inconsistent with the keyframe ctor that have an interpolator as parameter.
I prefer this version : KeyFramedValue do not know how to be interpolate, but just handles the values. Then linearInterpolate functions do the job.

Reply:
see answer above

Reply:
This is not resolved. KeyFramedValues are just an arrangement (possibly with constraints) of <time/value> pairs that only manage the storage and the coherency of the data. I'm not sure a keyframedvalue must hold an interpolator.